src/lib/provable/gadgets/twisted-curve.ts

```diff
@@ -0,0 +1,513 @@
+import { inverse, mod } from '../../../bindings/crypto/finite-field.js';
+import { Provable } from '../provable.js';
+import { assert } from './common.js';
+import { Field3, ForeignField, split } from './foreign-field.js';
+import { l2Mask } from './range-check.js';
+import { sha256 } from 'js-sha256';
+import { provable } from '../types/provable-derivers.js';
+import {
+  bigIntToBytes,
+  bytesToBigInt,
+} from '../../../bindings/crypto/bigint-helpers.js';
+import {
+  CurveTwisted,
+  GroupTwisted,
+  twistedAdd,
+  twistedDouble,
+} from '../../../bindings/crypto/elliptic-curve.js';
+import { assertPositiveInteger } from '../../../bindings/crypto/non-negative.js';
+import { sliceField3 } from './bit-slices.js';
+import { exists } from '../core/exists.js';
+import { arrayGetGeneric } from './elliptic-curve.js';
+
+// external API
+export { CurveTwisted };
+
+// internal API
+export { Point, initialAggregator, simpleMapToCurve, arrayGetGeneric };
+
+const CurveTwisted = {
+  add,
+  double,
+  negate,
+  assertOnCurve,
+  scale,
+  multiScalarMul,
+};
+
+/**
+ * Non-zero twisted elliptic curve point.
+ */
+type Point = { x: Field3; y: Field3 };
+type point = { x: bigint; y: bigint };
+
+const Point = {
+  from({ x, y }: point): Point {
+    return { x: Field3.from(x), y: Field3.from(y) };
+  },
+  toBigint({ x, y }: Point) {
+    let x_ = Field3.toBigint(x);
+    let y_ = Field3.toBigint(y);
+    return { x: x_, y: y_, infinity: x_ === 0n && y_ === 1n };
+  },
+  isConstant: (P: Point) => Provable.isConstant(Point, P),
+
+  /**
+   * Random point on the curve.
+   */
+  random(Curve: CurveTwisted) {
+    return Point.from(random(Curve));
+  },
+
+  provable: provable({ x: Field3, y: Field3 }),
+};
+
+function add(
+  p1: Point,
+  p2: Point,
+  Curve: { modulus: bigint; a: bigint; d: bigint }
+) {
+  let { x: x1, y: y1 } = p1;
+  let { x: x2, y: y2 } = p2;
+  let f = Curve.modulus;
+  let a = Curve.a;
+  let d = Curve.d;
+
+  // constant case
+  if (Point.isConstant(p1) && Point.isConstant(p2)) {
+    let p3 = twistedAdd(Point.toBigint(p1), Point.toBigint(p2), f, a, d);
+    return Point.from(p3);
+  }
+
+  assert(
+    Curve.modulus > l2Mask + 1n,
+    'Base field moduli smaller than 2^176 are not supported'
+  );
+
+  // witness and range-check denominators, x3, y3
+  let witnesses = exists(12, () => {
+    let [x1_, x2_, y1_, y2_] = Field3.toBigints(x1, x2, y1, y2);
+
+    // TODO: reuse code in twistedAdd to avoid recomputing these
```

Comment:
This would be a nice feature and much cleaner code, but apparently this wouldn't be a trivial refactor at all. Ideas for the future: inspiration from the interpreter structure in o1vm?